src/qibolab/platforms/multiqubit.py

```diff
@@ -45,13 +130,37 @@ def execute_pulse_sequence(self, sequence: PulseSequence, nshots=None):
             if "readout" in roles[name]:
                 if not instrument_pulses[name].is_empty:
                     if not instrument_pulses[name].ro_pulses.is_empty:
+                        if all([pulse.serial in changed for pulse in instrument_pulses[name].ro_pulses]):
+                            # FIXME: for precision sweep in resonator spectroscopy
+                            # change necessary to perform precision sweep
+                            # TODO: move this to instruments (ask Alvaro)
+                            # TODO: check if this will work with multiplex
+                            for sequencers in self.instruments[name]._sequencers.values():
+                                for sequencer in sequencers:
+                                    sequencer.pulses = instrument_pulses[name].ro_pulses
                         results = self.instruments[name].acquire()
                         existing_keys = set(acquisition_results.keys()) & set(results.keys())
                         for key, value in results.items():
                             if key in existing_keys:
                                 acquisition_results[key].update(value)
                             else:
                                 acquisition_results[key] = value
+
+        # change back the frequency of the pulses
+        for name in self.instruments:
+            roles[name] = self.settings["instruments"][name]["roles"]
+            if "readout" in roles[name]:
+                instrument_pulses[name] = sequence.get_channel_pulses(*self.instruments[name].channels)
```

Comment:
Never access the object you are iterating over. In case you need it, just iterate over items:
```py
for name, instr in self.instruments.items():
```
```suggestion
                instrument_pulses[name] = sequence.get_channel_pulses(*instr.channels)
```